### ISSUE TITLE
"{}" type issues when comapred to "Object"/"object"

### ISSUE THREAD

**Issue by dblachut-adb** [microsoft/TypeScript#48988]
# Bug Report

### 🔎 Search Terms

"object" | "{}"

### 🕗 Version & Regression Information

- This is the behavior in every version I tried, and I reviewed the FAQ for entries about object & {}.

### ⏯ Playground Link

[Playground link with relevant code](https://www.typescriptlang.org/play?jsx=0&ts=4.6.2&ssl=46&ssc=1&pln=1&pc=1#code/PTAEGIEYC5QdQBYEMAuoCWBnUAiA3gL46goCeADgKYD8oAKglqJpZQLbYoD2oARpaACuAOwAmXAMaC2lYSkqiAdACgyVUAEFQAXlCFlErsMxokABlhbdI0ZQBm6YQtAhQlAE7uu7g0ZOgkGE0dUGFBABtwlzAPLx9DY1MAJksQpOjQLgBrX0SAgGZU3QByBEpIngB3b3DRYujs3P8kABYivVA7Li5YYt4kd3qCBpyE5oBWdoAKAEodAD4hMXtHZxBG5VcAYS42cgGsIxIeHAB5XgArSgkUYjVKVQoBACEQ86ubprReC1BX62WDicogysW8Xz4QX+oQiUVcYPifm+KT+aQyjTG30KqJKZQqoGq7lq9XWoyRfDaOI6XR6oD6AyGIwhvEmVNmCyWtiBa2AG22u327kOwmOuC4l2utxIT0e6i2IXFHxQEIkv3lAK5qxB8M84MxoAkQXVMMioN1iLyEhRxvSOriKuxxtK5XCVRqdTN9v1EkpxrwnW6vX6g1Aw1JKtZxvZ2kWNhWwKZmzA4BRiFImQlN1A4XQ8ncSCiyGwSFwhDuTwJZRFjk6SCzKEY2EwCC4ERB-ACIrOmduwBwislOGoEIAGiFCBkR7B9PdQAAREL3Lh2UBj1wL3T6JMQbFp6VUZehSiVUDvSWgIsBUtEfcCSpVjAiux1tANpjN1u1PgCJBds83HA+wHADh31ABNEInBPf8UHZVwwNgGDZQEABRRcnkPNcwDQzcCGUbdwEpPd8BvWcJF-b8AkwTBBAAcwTbhQDYbwBGEAYvBPWd+x7HBFD45RHDzZ8JAEAAxAAeOg3AAD3kMRsGAtBcMWPBlFAdTQHILxyFgOgAG5lDw5DQAAOWkfh3FOSye0g8yPFAAAfDMlQM1w+MUe90HCAQGwEfhXRPWj0AAN0oYsRQRYdBI8YSBAAcUkmS5NEBSbN0My2AsqyYJUtSNK0rgdPoAy8KAA)

### 💻 Code
1. 
```ts
type A = {}
const a0: A = undefined // error
const a1: A = null // error
const a2: A = 2 // ok
const a3: A = 'hello world' //ok
const a4: A = { foo: 'bar' } //ok
const a5: A = () => undefined //ok
```
2. 
```ts
const X = {} // X: {}
type D = typeof X // D = {}
```
3.
```ts
const Y = new Object() // Y: Object
type E = typeof X // E = {}
```
4.
```ts
interface F<T extends object = {}> {
    prop: T;
}
```

### 🙁 Actual behavior

I have found at least few issues with `{}` type, all presented in [playground](https://www.typescriptlang.org/play?jsx=0&ts=4.6.2&ssl=46&ssc=1&pln=1&pc=1#code/PTAEGIEYC5QdQBYEMAuoCWBnUAiA3gL46goCeADgKYD8oAKglqJpZQLbYoD2oARpaACuAOwAmXAMaC2lYSkqiAdACgyVUAEFQAXlCFlErsMxokABlhbdI0ZQBm6YQtAhQlAE7uu7g0ZOgkGE0dUGFBABtwlzAPLx9DY1MAJksQpOjQLgBrX0SAgGZU3QByBEpIngB3b3DRYujs3P8kABYivVA7Li5YYt4kd3qCBpyE5oBWdoAKAEodAD4hMXtHZxBG5VcAYS42cgGsIxIeHAB5XgArSgkUYjVKVQoBACEQ86ubprReC1BX62WDicogysW8Xz4QX+oQiUVcYPifm+KT+aQyjTG30KqJKZQqoGq7lq9XWoyRfDaOI6XR6oD6AyGIwhvEmVNmCyWtiBa2AG22u327kOwmOuC4l2utxIT0e6i2IXFHxQEIkv3lAK5qxB8M84MxoAkQXVMMioN1iLyEhRxvSOriKuxxtK5XCVRqdTN9v1EkpxrwnW6vX6g1Aw1JKtZxvZ2kWNhWwKZmzA4BRiFImQlN1A4XQ8ncSCiyGwSFwhDuTwJZRFjk6SCzKEY2EwCC4ERB-ACIrOmduwBwislOGoEIAGiFCBkR7B9PdQAAREL3Lh2UBj1wL3T6JMQbFp6VUZehSiVUDvSWgIsBUtEfcCSpVjAiux1tANpjN1u1PgCJBds83HA+wHADh31ABNEInBPf8UHZVwwNgGDZQEABRRcnkPNcwDQzcCGUbdwEpPd8BvWcJF-b8AkwTBBAAcwTbhQDYbwBGEAYvBPWd+x7HBFD45RHDzZ8JAEAAxAAeOg3AAD3kMRsGAtBcMWPBlFAdTQHILxyFgOgAG5lDw5DQAAOWkfh3FOSye0g8yPFAAAfDMlQM1w+MUe90HCAQGwEfhXRPWj0AAN0oYsRQRYdBI8YSBAAcUkmS5NEBSbN0My2AsqyYJUtSNK0rgdPoAy8KAA) and code samples above:
1. Is undocumented (or I couldn't find it in docs) and confusing - looks like object literal but can have primitives assigned.
2. Object literal has a `{}` type assumed while it should rather be `Object`/`object`.
3. `new Object` has a `{}` type assumed while it should rather be `Object`/`object`.
4. `{}` type can be assigned to more narrow `object` type, which should not be possible.

### 🙂 Expected behavior

`{}` type is documented, cannot be assigned to more narrow `object` type and is not assumed when used with `typeof <some_object>`.

**Comment by dblachut-adb** [comment]
Okay so `{}` is an empty interface and that is the root of all evil. :wink:
Makes sense, but can also be a source of errors, for example:
```ts
const X = {} // X: {}
const variableWithSameTypeAsX: typeof X = 1;

// typescript allows compaison of "number" and "object" which does not make sense
if (variableWithSameTypeAsX == X) {

}
```
[Playground link](https://www.typescriptlang.org/play?jsx=0&ts=4.6.2&ssl=7&ssc=2&pln=1&pc=1#code/MYewdgzgLgBAGjAvDA3gXxgek-AXKtAKFElgDcBDAJwEsKAjAGwFMB1GqACwGUKBbZgBUAngAdmAQQhx8UMcxAAzeEhgBGANyFC2GHPERgtUbAqNGIAO4QYoPqIo0I4GEpgAiMAFc+9ZlXcYCjAAEw8QegArZmAoQMtOGmBOGBCQZhswEFg+CgBrZhgIZkhmQhplAApKWgYWdi5eARFxKQREZDgASlRtNCA).

In other words: IMHO primitives shall not be assigned to empty interface.

**Comment by MartinJohns** [comment]
It's intentional. You basically shouldn't be using `{}`, unless you understood it and have a good reason for it. It means "any type, except `undefined` and `null`."

**Comment by dblachut-adb** [comment]
Maybe such distinction should exist for the reasons I mentioned - you cannot have any properties added to a primitive.

So the only possible assignment of primitive to an interface (except empty one which accepts everything) is when this interface have properties defined in it's boxed prototype.

Boxed prototypes are well-defined so maybe there is a way to restrict primitive to interface assignment while still allowing such code:
```ts
interface L { length: number }
const X: L = "str";
```

**Comment by MartinJohns** [comment]
I'm sure the TypeScript team has thought about this the numerous times this issue came up. Like Ryan mentioned, it has implications on subtype reduction.

But I fail to see the general issue this causes. Can you provide w real example? Your earlier example where you mentioned the comparison between number and object is fine, because your variable typed `{}` can either store a number or an object.

**Comment by Josh-Cena** [comment]
I don't know if `{}` being different from all other object types is a design decision or an oversight, but it's more or less a feature now: people use it to represent any non-nullish value. There's no other type that represents the same thing.

**Comment by MartinJohns** [comment]
I don't see how it's being different. There only seems to be the wrong assuming that the **type** `{}` means it's an object. Neither does `interface` or `type` mean you're dealing with an object.

**Comment by RyanCavanaugh** [comment]
You can't make the primitives unassignable to `{ }` without breaking valuable symmetry. It would render at least one of these bullets incorrect:

Here's a proof of why `string` is `{ }`

 * A value `x` has a property `k` if `x.k` is a legal expression
 * `"hello world".length` is a legal expression
   * Therefore `string` has a `length` property
 * Things with a `length` property are subtypes of `{ length: number }`
   * Therefore, `string` is a subtype of `{ length: number }`
 * For any two type definitions `{ a: A, b: B }` and `{ a: A }`, the former is a subtype of the latter
   * This is inductive all the way to zero: `{ a: A }` is a subtype of `{ }`
 * Therefore, `string` is a subtype of `{ }`

You can break this proof at any single step, but it introduces really bad assymmetries without fixing any other problem than "some people misapprehended that `{ }` is a synonym of `object`", and it's impossible to remove all misapprehensions from a system.

**Comment by dblachut-adb** [comment]
Question related: can you have a type in typescript which represents object without any properties? So that it would be possible to assign only empty object to that type.

**Comment by MartinJohns** [comment]
@dblachut-adb No, currently you can't define such a type in TypeScript. This would require #12936.
